scipy/special/xsf/zeta.h

```diff
@@ -1,10 +1,279 @@
+/* Complex riemann-zeta function implementation based on Python implementation
+ * written by Matt Haberland (@mdhaber) in:
+ * https://colab.research.google.com/drive/1zMDSAJlXCLRqMMtJ0e9nDGjQ8iZCnmn5?usp=sharing
+ */
+
 #pragma once
 
+#include "config.h"
+#include "error.h"
+
+#include "cephes/const.h"
 #include "cephes/zeta.h"
 #include "cephes/zetac.h"
 
 namespace xsf {
 
+namespace detail {
+
+    /* Log of absolute value of expansion coefficients for Euler-Maclaurin
+     * summation formula.
+     * log(|B2k / (2k)!|) */
+    constexpr double zeta_em_log_abs_coeff_lookup[] = {
+	-2.4849066497880004,
+	-6.579251212010101,
+	-10.31692083029347,
+	-14.005800284407405,
+	-17.68462940266784,
+	-21.361131560073222,
+	-25.037070502911423,
+	-28.712870599846948,
+	-32.388636197522295,
+	-36.06439319366539,
+	-39.740148041995184,
+	-43.41590235365616,
+	-47.091656531181485,
+	-50.76741067517639,
+	-54.44316481078909,
+	-58.11891894430628,
+	-61.79467307729959,
+	-65.47042721016194,
+	-69.14618134299154,
+	-72.82193547581296,
+	-76.49768960863234,
+	-80.1734437414512,
+	-83.84919787426993,
+	-87.52495200708863,
+	-91.20070613990733,
+	-94.87646027272602,
+	-98.55221440554472,
+	-102.2279685383634,
+	-105.9037226711821,
+	-109.57947680400078,
+	-113.25523093681947,
+	-116.93098506963817,
+	-120.60673920245685,
+	-124.28249333527555,
+	-127.95824746809424,
+	-131.63400160091294,
+	-135.30975573373163,
+	-138.9855098665503,
+	-142.661263999369,
+	-146.3370181321877,
+	-150.0127722650064,
+	-153.6885263978251,
+	-157.36428053064375,
+	-161.04003466346245,
+	-164.71578879628115,
+	-168.39154292909984,
+	-172.06729706191854,
+	-175.74305119473723,
+	-179.4188053275559,
+	-183.0945594603746
+    };
+
+    // Complex log of expansion coefficients for Euler-Maclaurin summation formula.
+    XSF_HOST_DEVICE inline std::complex<double> zeta_em_log_coeff(std::size_t n) {
+	std::complex<double> J(0.0, 1.0);
+	std::complex<double> result;
+	if (n < 50) {
+	    result = zeta_em_log_abs_coeff_lookup[n];
+	} else {
+	    // asymptotic formula
+	    result = std::log(2.0) - 2.0*(n + 1)*std::log(2*M_PI);
```

Comment:
By the asymptotic approximation for $|B_2n|$ we have

$$\frac{|B_{2n}|}{2n!} \sim \frac{4\sqrt{\pi n}\left(\frac{n}{\pi e}\right)^{2n}}{2n!}$$

and then by Stirling's formula this is

$$\sim \frac{4\sqrt{\pi n}\left(\frac{n}{\pi e}\right)^{2n}}{\sqrt{4\pi n}\left(\frac{2n}{e}\right)^{2n}}$$

we can then do some cancellations to get

$$\sim \frac{2}{\left(2\pi\right)^{2n}}$$

and then take logs. The other thing that's confusing here is that `zeta_em_log_coeff` is zero indexed instead of 1 indexed, so `zeta_em_log_coeff(i)` gives $\log\left|\frac{B_{2(i + 1)}}{\left(2(i+1)\right)!}\right|$

the reason i did that is to make it straightforward to have it so that the function will work for any `size_t` `n`

Reply:
Yeah, this is much easier with the [formula you linked](https://dlmf.nist.gov/24.11#E1) rather than the one I used.
24.11.1 gives:
```math
|B_{2n}| \sim \frac{2(2n)!}{(2\pi)^{2n}}
```
directly. Dividing by $(2n)!$ just removes that term from the numerator.
```math
|B_{2n}| \sim \frac{2(}{(2\pi)^{2n}}
```